### PR TITLE
[SPARK-24339][SQL] Prunes the unused columns from child of ScriptTransformation

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

Modify the strategy in ColumnPruning to add a Project between ScriptTransformation and its child, this strategy can reduce the scan time especially in the scenario of the table has many columns.

## How was this patch tested?

Add UT in ColumnPruningSuite and ScriptTransformationSuite.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
@gatorsmile @maropu This is the follow up PR for #21447, please have a look when you have time, thanks.

**Comment by maropu** [comment]
Thanks for the work! I’ll review this in hours.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -450,13 +450,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
-    // Prunes the unused columns from child of Aggregate/Expand/Generate
+    // Prunes the unused columns from child of Aggregate/Expand/Generate/ScriptTransformation
     case a @ Aggregate(_, _, child) if (child.outputSet -- a.references).nonEmpty =>
       a.copy(child = prunedChild(child, a.references))
     case f @ FlatMapGroupsInPandas(_, _, _, child) if (child.outputSet -- f.references).nonEmpty =>
       f.copy(child = prunedChild(child, f.references))
     case e @ Expand(_, _, child) if (child.outputSet -- e.references).nonEmpty =>
       e.copy(child = prunedChild(child, e.references))
+    case s @ ScriptTransformation(_, _, _, child, _)
+      if (child.outputSet -- s.references).nonEmpty =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: two more spaces before `if`.

**Comment by gatorsmile** [comment]
@xuanyuanking Could you add an end-to-end test into ScriptTransformationSuite to verify the results?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thanks, fix in 2cf131f.

**Comment by xuanyuanking** [comment]
@gatorsmile Thanks for your advice, added ut in ScriptTransformationSuite.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by xuanyuanking** [comment]
Thanks for reviewing.

**Pull request closed without merging**